File: src/test/java/kitchenpos/eatinorder/domain/model/todo/EatInOrderTest.java
package kitchenpos.eatinorder.domain.model.todo;

import org.assertj.core.api.ThrowableAssert;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertAll;

class EatInOrderTest {
    @DisplayName("EeaInOrder를 생성한다.")
    @Test
    void create() {
        // given
        final UUID id = UUID.randomUUID();
        final LocalDateTime orderDateTime = LocalDateTime.now();
        final UUID orderTableId = UUID.randomUUID();
        final EatInOrderLineItem eatInOrderLineItem = EatInOrderLineItem.of(1L, UUID.randomUUID(), 1L, 1L, true);
        final List<EatInOrderLineItem> eatInOrderLineItems = List.of(eatInOrderLineItem);

        // when
        final EatInOrder eatInOrder = EatInOrder.create(id, orderDateTime, eatInOrderLineItems, orderTableId, orderTableId1 -> {});

        // then
        assertAll(
                () -> assertThat(eatInOrder).isNotNull(),
                () -> assertThat(eatInOrder.getOrderTableId()).isEqualTo(orderTableId),
                () -> assertThat(eatInOrder.getOrderDateTime()).isEqualTo(orderDateTime),
                () -> assertThat(eatInOrder.getStatus()).isEqualTo(EatInOrderStatus.WAITING),
                () -> assertThat(eatInOrder.getLineItems()).containsExactly(eatInOrderLineItem)
        );
    }

    @DisplayName("EeaInOrder를 생성할 때 주문 테이블이 사용 중이면 예외를 던진다.")
    @Test
    void createWithOccupiedOrderTable() {
        // given
        final UUID id = UUID.randomUUID();
        final LocalDateTime orderDateTime = LocalDateTime.now();
        final UUID orderTableId = UUID.randomUUID();
        final EatInOrderLineItem eatInOrderLineItem = EatInOrderLineItem.of(1L, UUID.randomUUID(), 1L, 1L, true);
        final List<EatInOrderLineItem> eatInOrderLineItems = List.of(eatInOrderLineItem);

        // when
        final Throwable thrown = catchThrowable(() -> EatInOrder.create(id, orderDateTime, eatInOrderLineItems, orderTableId, orderTableId1 -> {
            throw new IllegalStateException();
        }));

        // then
        assertThat(thrown).isInstanceOf(IllegalStateException.class);
    }

    @DisplayName("EeaInOrder를 수락한다.")
    @Test
    void accept() {
        // given
        final EatInOrder eatInOrder = createEatInOrder();

        // when
        eatInOrder.accept();

        // then
        assertThat(eatInOrder.getStatus()).isEqualTo(EatInOrderStatus.ACCEPTED);
    }

    @DisplayName("주문 상태가 '대기' 상태인 경우에만 수락할 수 있다.")
    @ValueSource(strings = {"ACCEPTED", "SERVED", "COMPLETED"})
    @ParameterizedTest
    void acceptWhenWaiting(EatInOrderStatus status) {
        // given
        final EatInOrder eatInOrder = createEatInOrder(status);

        // when
        ThrowableAssert.ThrowingCallable accept = eatInOrder::accept;

        // then
        assertThatThrownBy(accept)
                .isInstanceOf(IllegalStateException.class);
    }

    @DisplayName("EeaInOrder를 제공한다.")
    @Test
    void serve() {
        // given
        final EatInOrder eatInOrder = createEatInOrder(EatInOrderStatus.ACCEPTED);

        // when
        eatInOrder.serve();

        // then
        assertThat(eatInOrder.getStatus()).isEqualTo(EatInOrderStatus.SERVED);
    }

    @DisplayName("주문 상태가 '수락' 상태인 경우에만 제공할 수 있다.")
    @ValueSource(strings = {"WAITING", "SERVED", "COMPLETED"})
    @ParameterizedTest
    void serveWhenAccepted(EatInOrderStatus status) {
        // given
        final EatInOrder eatInOrder = createEatInOrder(status);

        // when
        ThrowableAssert.ThrowingCallable serve = eatInOrder::serve;

        // then
        assertThatThrownBy(serve)
                .isInstanceOf(IllegalStateException.class);
    }

    @DisplayName("EeaInOrder를 완료한다.")
    @Test
    void complete() {
        // given
        final EatInOrder eatInOrder = createEatInOrder(EatInOrderStatus.SERVED);

        // when
        eatInOrder.complete();

        // then
        assertThat(eatInOrder.getStatus()).isEqualTo(EatInOrderStatus.COMPLETED);
    }

    @DisplayName("주문 상태가 '제공' 상태인 경우에만 완료할 수 있다.")
    @ValueSource(strings = {"WAITING", "ACCEPTED", "COMPLETED"})
    @ParameterizedTest
    void completeWhenServed(EatInOrderStatus status) {
        // given
        final EatInOrder eatInOrder = createEatInOrder(status);

        // when
        ThrowableAssert.ThrowingCallable complete = eatInOrder::complete;

        // then
        assertThatThrownBy(complete)
                .isInstanceOf(IllegalStateException.class);
    }

    private static EatInOrder createEatInOrder() {
        return createEatInOrder(EatInOrderStatus.WAITING);
    }

    private static EatInOrder createEatInOrder(EatInOrderStatus status) {
        final UUID id = UUID.randomUUID();
        final LocalDateTime orderDateTime = LocalDateTime.now();
        final UUID orderTableId = UUID.randomUUID();
        final EatInOrderLineItem eatInOrderLineItem = EatInOrderLineItem.of(1L, UUID.randomUUID(), 1L, 1L, true);
        final List<EatInOrderLineItem> eatInOrderLineItems = List.of(eatInOrderLineItem);
        return EatInOrder.create(id, orderDateTime, eatInOrderLineItems, orderTableId, status);
    }
}

File: src/test/kotlin/kitchenpos/tobe/domain/eatinorder/domain/EatInOrderTest.kt
package kitchenpos.tobe.domain.eatinorder.domain

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldNotBe
import kitchenpos.tobe.eatinorder.domain.EatInOrderStatus
import kitchenpos.tobe.eatinorder.domain.entity.EatInOrder
import tobe.domain.EatInOrderFixtures.createEatInOrder
import tobe.domain.OrderLineItemV2Fixtures.createOrderLineItemV2
import tobe.domain.OrderTableV2Fixtures.createOrderTableV2
import java.time.LocalDateTime

class EatInOrderTest : DescribeSpec() {
    init {
        describe("EatInOrder 클래스의") {
            describe("from 메소드는") {
                context("적절한 요청이 주어졌을때") {
                    it("매장 내 식사 주문 정보를 반환한다") {
                        val eatInOrder =
                            EatInOrder.from(
                                orderDateTime = LocalDateTime.now(),
                                orderLineItems = listOf(createOrderLineItemV2()),
                                orderTable = createOrderTableV2(),
                            )

                        eatInOrder shouldNotBe null
                    }
                }
            }

            describe("accept 메소드는") {
                context("대기 상태가 아니라면") {
                    val eatInOrder =
                        createEatInOrder().also {
                            it.status = EatInOrderStatus.SERVED
                        }
                    it("예외를 던진다") {
                        shouldThrow<IllegalArgumentException> { eatInOrder.accept() }
                    }
                }
            }

            describe("serve 메소드는") {
                context("접수됨 상태가 아니라면") {
                    val eatInOrder =
                        createEatInOrder().also {
                            it.status = EatInOrderStatus.SERVED
                        }
                    it("예외를 던진다") {
                        val exception = shouldThrow<IllegalArgumentException> { eatInOrder.serve() }
                        exception.message?.contains("접수됨 상태의 주문만 제공할 수 있습니다.")
                    }
                }
            }

            describe("complete 메소드는") {
                context("제공됨 상태가 아니라면") {
                    val eatInOrder =
                        createEatInOrder().also {
                            it.status = EatInOrderStatus.WAITING
                        }
                    it("예외를 던진다") {
                        val exception = shouldThrow<IllegalArgumentException> { eatInOrder.complete() }
                        exception.message?.contains("제공됨 상태의 주문만 수락할 수 있습니다.")
                    }
                }
            }
        }
    }
}
